crates/core/src/client.rs

```diff
@@ -0,0 +1,253 @@
+use ethers::types::{Address, SyncingStatus};
+use eyre::{eyre, Result};
+use helios::client::Client;
+#[cfg(target_arch = "wasm32")]
+use helios::prelude::ConfigDB;
+use helios::prelude::Database;
+#[cfg(not(target_arch = "wasm32"))]
+use helios::prelude::FileDB;
+use helios::types::BlockTag;
+use serde_json::json;
+use starknet::core::types::{BlockId, BlockTag as SnBlockTag, FieldElement, MaybePendingBlockWithTxHashes};
+use starknet::providers::jsonrpc::{HttpTransport, JsonRpcClient};
+use starknet::providers::Provider;
+use tracing::{debug, info, warn};
+
+use crate::config::Config;
+use crate::storage_proofs::types::StorageProofResponse;
+use crate::storage_proofs::StorageProof;
+use crate::utils::*;
+use crate::CoreError;
+
+abigen!(
+    StarknetCoreContracts,
+    r#"[
+        function stateRoot() external view returns (uint256)
+        function stateBlockNumber() external view returns (int256)
+        function stateBlockHash() external view returns (uint256)
+        function l1ToL2Messages(bytes32 msgHash) external view returns (uint256)
+        function l2ToL1Messages(bytes32 msgHash) external view returns (uint256)
+        function l1ToL2MessageNonce() public view returns (uint256)
+        function l1ToL2MessageCancellations(bytes32 msgHash) external view returns (uint256) 
+    ]"#,
+    event_derives(serde::Deserialize, serde::Serialize)
+);
+
+#[derive(Clone, Debug)]
+pub struct NodeData {
+    pub l1_state_root: FieldElement,
+    pub l1_block_num: u64,
+    pub sync_root: FieldElement,
+    pub sync_block_num: u64,
+}
+
+impl NodeData {
+    pub fn new() -> Self {
+        NodeData {
+            l1_state_root: FieldElement::ZERO,
+            l1_block_num: 0,
+            sync_root: FieldElement::ZERO,
+            sync_block_num: 0,
+        }
+    }
+}
+
+impl Default for NodeData {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct BeerusClient {
+    /// Ethereum light client
+    #[cfg(not(target_arch = "wasm32"))]
+    pub helios_client: Arc<Client<FileDB>>,
+    #[cfg(target_arch = "wasm32")]
+    pub helios_client: Arc<Client<ConfigDB>>,
+    /// StarkNet light client
+    pub starknet_client: JsonRpcClient<HttpTransport>,
+    /// Proof client
+    pub proof_addr: String,
+    /// Poll interval
+    pub poll_secs: u64,
+    /// StarkNet core contract address
+    pub core_contract_addr: Address,
+    /// Payload data
+    pub node: Arc<RwLock<NodeData>>,
+    /// Beerus client config
+    pub config: Config,
+}
+
+impl BeerusClient {
+    /// Create a new Beerus Light Client service.
+    pub async fn new(config: Config) -> Self {
+        #[cfg(not(target_arch = "wasm32"))]
+        let mut helios_client: Client<FileDB> = config.to_helios_client().await;
+        #[cfg(target_arch = "wasm32")]
+        let mut helios_client: Client<ConfigDB> = config.to_helios_client().await;
+
+        helios_client.start().await.expect("could not init helios client");
```

Comment:
helios doesn't fail if the endpoints are incorrect like it did; So that now beerus throws panics from a thread in a loop. Maybe consider adding some request and exit right away in case of a bad configuration

Reply:
Added a short curcuit before we hit the loop:
```
        if let Err(err) = sn_state_root_inner(&helios_client, config.get_core_contract_address()).await {
            panic!("execution client err: \n\n{err:#?}");
        }
```